File: Problems/FindTheMissingSmallPositiveNumber.js
/*
 * Complete the 'findSmallestMissingPositive' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts INTEGER_ARRAY orderNumbers as parameter.
 */

function findSmallestMissingPositive(orderNumbers) {
    // Write your code here
    let i = 0;
    let j = orderNumbers.length - 1;
    let ans = 0;
    
    while (i < j) {
        if (orderNumbers[i] - 1 === i) {
            i += 1;
        } else if (orderNumbers[i] <= j && orderNumbers[i] > 0) {
            const number = orderNumbers[i] - 1;
            [orderNumbers[i], orderNumbers[number]] = [orderNumbers[number], orderNumbers[i]];
            if (orderNumbers[i] === orderNumbers[number]) {
                [orderNumbers[j], orderNumbers[i]] = [orderNumbers[i], orderNumbers[j]];
                j--;
            }
        } else {
            [orderNumbers[j], orderNumbers[i]] = [orderNumbers[i], orderNumbers[j]];
            j--;
        }
        
    }
    
    for (let i = 0; i < orderNumbers.length; i++) {
        if (orderNumbers[i] - 1 === i) {
            ans = orderNumbers[i]
        } else {
            break;
        } 
    }
    
    
    return ans + 1;

}

console.log(findSmallestMissingPositive([5,4,3,2,1,6]));      // expected 2
console.log(findSmallestMissingPositive([3, 4, -1, 1]));  // expected 2
console.log(findSmallestMissingPositive([7, 8, 9, 11]));  // expected 1

File: Problems/FindTheMissingSmallPositiveNumber.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FindTheMissingSmallPositiveNumber {
    public static int findSmallestMissingPositive(List<Integer> orderNumbers) {
        if (orderNumbers.isEmpty()) {
            return 1;
        }

        int len = orderNumbers.size();

        for (int i = 0; i < len; i++) {
            while (true) {
                int val = orderNumbers.get(i);

                if (!(val > 0 && val <= len && val != orderNumbers.get(val - 1))) {
                    break;
                }

                int tmp = orderNumbers.get(val - 1);
                orderNumbers.set(val - 1, val);
                orderNumbers.set(i, tmp);
            }
        }

        for (int i = 0; i < len; i++) {
            if (orderNumbers.get(i) != i + 1) {
                return i + 1;
            }
        }

        return len + 1;
    }

    public static void main(String[] args) {
        System.out.println(findSmallestMissingPositive(new ArrayList<>(Arrays.asList(-3,-2,-1,0,1,-4)))); // expected 1
        System.out.println(findSmallestMissingPositive(new ArrayList<>(Arrays.asList(2, 2, 1)))); // expected 3
        System.out.println(findSmallestMissingPositive(new ArrayList<>(Arrays.asList(3, 4, -1, 1)))); // expected 2
        System.out.println(findSmallestMissingPositive(new ArrayList<>(Arrays.asList(7, 8, 9, 11)))); // expected 1
    }
}
